spinoso-symbol/src/inspect.rs

```diff
@@ -0,0 +1,734 @@
+        }
+        if self.open_quote {
+            self.open_quote = false;
+            return Some('"');
+        }
+        for cell in &mut self.next {
+            let mut done = false;
+            let mut next = None;
+            if let Some(ref mut lit) = cell {
+                next = lit.next();
+                done = next.is_none();
+            }
+            if done {
+                *cell = None;
+            }
+            if next.is_some() {
+                return next;
+            }
+        }
+        if let Some((start, end, ch)) = self.iter.next() {
+            if ch == '\u{FFFD}' {
+                let mut next = None::<char>;
+                let slice = &self.string[start..end];
+                let iter = slice.iter().zip(self.next.iter_mut()).enumerate();
+                for (idx, (&byte, cell)) in iter {
+                    let mut lit = Literal::from(byte);
+                    if idx == 0 {
+                        next = lit.0.next();
+                    }
+                    *cell = Some(lit);
+                }
+                return next;
+            } else {
+                return Some(ch);
+            }
+        }
+        if self.close_quote {
+            self.close_quote = false;
+            return Some('"');
+        }
+        None
+    }
+}
+
+impl<'a> DoubleEndedIterator for State<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.close_quote {
+            self.close_quote = false;
+            return Some('"');
+        }
+        for cell in self.next.iter_mut().rev() {
+            let mut done = false;
+            let mut next = None;
+            if let Some(ref mut lit) = cell {
+                next = lit.next_back();
+                done = next.is_none();
+            }
+            if done {
+                *cell = None;
+            }
+            if next.is_some() {
+                return next;
+            }
+        }
+        if let Some((start, end, ch)) = self.iter.next_back() {
+            if ch == '\u{FFFD}' {
+                let mut next = None::<char>;
+                let slice = &self.string[start..end];
+                let iter = slice.iter().zip(self.next.iter_mut()).rev().enumerate();
+                for (idx, (&byte, cell)) in iter {
+                    let mut lit = Literal::from(byte);
+                    if idx == 0 {
+                        next = lit.0.next_back();
+                    }
+                    *cell = Some(lit);
+                }
+                return next;
+            } else {
+                return Some(ch);
+            }
+        }
+        if self.open_quote {
+            self.open_quote = false;
+            return Some('"');
+        }
+        if self.sigil {
+            self.sigil = false;
+            return Some(':');
+        }
+        None
+    }
+}
+
+impl<'a> FusedIterator for State<'a> {}
+
+#[must_use = "Iterator"]
+#[derive(Debug, Clone)]
+struct Literal(Chars<'static>);
```

Comment:
Breaking down parts of the `State` iterator into smaller iterators helped a lot with making the logic more readable.

This is also a nice hack to get a `'static` iterator of `\xFF` style escapes for single bytes.